File: tsms-frontend/src/app/services/stock.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Stock } from '../models/models';

@Injectable({ providedIn: 'root' })
export class StockService {
  private baseUrl = 'http://localhost:8080/api/stock';

  constructor(private http: HttpClient) {}

  getAll(): Observable<Stock[]> {
    return this.http.get<Stock[]>(this.baseUrl);
  }

  getById(id: number): Observable<Stock> {
    return this.http.get<Stock>(`${this.baseUrl}/${id}`);
  }

  search(item: string): Observable<Stock[]> {
    return this.http.get<Stock[]>(`${this.baseUrl}/search?item=${item}`);
  }

  getByBrand(brandId: number): Observable<Stock[]> {
    return this.http.get<Stock[]>(`${this.baseUrl}/brand/${brandId}`);
  }

  getByStatus(status: string): Observable<Stock[]> {
    return this.http.get<Stock[]>(`${this.baseUrl}/status/${status}`);
  }

  create(stock: Stock): Observable<Stock> {
    return this.http.post<Stock>(this.baseUrl, stock);
  }

  update(id: number, stock: Stock): Observable<Stock> {
    return this.http.put<Stock>(`${this.baseUrl}/${id}`, stock);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: tsms-backend/src/main/java/com/tsms/service/StockService.java
package com.tsms.service;

import com.tsms.entity.Stock;
import com.tsms.repository.StockRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class StockService {

    private final StockRepository repository;

    public List<Stock> getAll() {
        return repository.findAll();
    }

    public Stock getById(Long id) {
        return repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Stock item not found with id: " + id));
    }

    public List<Stock> searchByItem(String stockItem) {
        return repository.findByStockItemContainingIgnoreCase(stockItem);
    }

    public List<Stock> getByBrand(Long brandId) {
        return repository.findByBrandId(brandId);
    }

    public List<Stock> getByStatus(String status) {
        return repository.findByStatus(status);
    }

    public Stock create(Stock stock) {
        return repository.save(stock);
    }

    public Stock update(Long id, Stock stock) {
        Stock existing = getById(id);
        existing.setStockItem(stock.getStockItem());
        existing.setBrand(stock.getBrand());
        existing.setSiNumber(stock.getSiNumber());
        existing.setStatus(stock.getStatus());
        return repository.save(existing);
    }

    public void delete(Long id) {
        repository.deleteById(id);
    }
}
